dcicutils/qa_checkers.py

```diff
@@ -0,0 +1,372 @@
+import glob
+import io
+import os
+import re
+import toml
+import warnings
+
+from collections import defaultdict
+from typing import Optional, List
+from .lang_utils import conjoined_list, n_of, there_are
+from .misc_utils import PRINT, remove_prefix, remove_suffix, getattr_customized, CustomizableProperty
+
+
+QA_EXCEPTION_PATTERN = re.compile(r"[#].*\b[N][O][Q][A]\b", re.IGNORECASE)
+
+
+def find_uses(*, where, patterns, recursive=False):
+    """
+    In the files specified by where (a glob pattern), finds uses of pattern (a regular expression).
+
+    :param where: a glob pattern
+    :param patterns: a dictionary mapping problem summaries to regular expressions
+    :param recursive: a boolean saying whether to look recursively [UNIMPLEMENTED]
+    """
+
+    assert not recursive, "The recursive option to find_uses is not yet implemented."
+
+    checks = []
+    for summary, pattern in patterns.items():
+        checks.append((re.compile(pattern), summary))
+    uses = defaultdict(lambda: [])
+    files = glob.glob(where)
+    for file in files:
```

Comment:
Pretty sure you can implement the recursive option by just passing `recursive=True` to `glob.glob`?

Reply:
I've fixed this and made other changes that it would be good if you could review. They're not really anything controversial, but they're of enough substance it counts as new work.

With the new code, I put up a beta and in cgap-portal I'm able to do:
```
@pytest.mark.static
def test_utils_debugging_artifacts_pdb():
    checker = DebuggingArtifactChecker(sources_subdir="src/encoded",
                                       skip_files="(tests/data)",
                                       filter_patterns=['pdb'])
    checker.check_for_debugging_patterns()


@pytest.mark.static
def test_utils_debugging_artifacts_print():
    checker = DebuggingArtifactChecker(sources_subdir="src/encoded",
                                       skip_files="encoded/(commands|tests)/",
                                       filter_patterns=['print'],
                                       if_used='warning')
    checker.check_for_debugging_patterns()
```
and get back:
```
========== test session starts ==========
platform darwin -- Python 3.8.13, pytest-3.10.1, py-1.11.0, pluggy-1.0.0 -- /Users/kentpitman/py/cgap-portal9/cg9env/bin/python
cachedir: .pytest_cache
rootdir: /Users/kentpitman/py/cgap-portal9, inifile: pytest.ini
plugins: cov-2.9.0, instafail-0.4.2, forked-1.4.0, mock-3.2.0, flaky-3.7.0, exact-fixtures-0.3, timeout-1.4.2, xdist-1.27.0
collected 1948 items / 1945 deselected                                                                                                                                                  

src/encoded/tests/test_misc.py::test_changelog_consistency PASSED           [ 33%]
src/encoded/tests/test_misc.py::test_utils_debugging_artifacts_pdb PASSED   [ 66%]
src/encoded/tests/test_misc.py::test_utils_debugging_artifacts_print PASSED [100%]

========== warnings summary ==========
src/encoded/tests/test_misc.py::test_utils_debugging_artifacts_print
  /Users/kentpitman/py/cgap-portal9/cg9env/lib/python3.8/site-packages/dcicutils/qa_checkers.py:356: SyntaxWarning: 24 problems detected:
   In /Users/kentpitman/py/cgap-portal9/src/encoded/visualization.py, 1 call to print.
   In /Users/kentpitman/py/cgap-portal9/src/encoded/authentication.py, 2 calls to print.
   In /Users/kentpitman/py/cgap-portal9/src/encoded/loadxl.py, 10 calls to print.
   In /Users/kentpitman/py/cgap-portal9/src/encoded/verifier.py, 3 calls to print.
   In /Users/kentpitman/py/cgap-portal9/src/encoded/types/ingestion.py, 4 calls to print.
   In /Users/kentpitman/py/cgap-portal9/src/encoded/types/file.py, 2 calls to print.
   In /Users/kentpitman/py/cgap-portal9/src/encoded/types/family.py, 1 call to print.
   In /Users/kentpitman/py/cgap-portal9/src/encoded/types/workflow.py, 1 call to print.
    confirm_no_uses(where=self.sources_pattern,

-- Docs: https://docs.pytest.org/en/latest/warnings.html
========== 3 passed, 1945 deselected, 1 warnings in 7.01 seconds ==========

```